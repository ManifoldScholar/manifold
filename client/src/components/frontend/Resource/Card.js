import React, { Component, PropTypes } from 'react';
import { Link, withRouter } from 'react-router-dom';
import classNames from 'classnames';
import FormattedDate from 'components/global/FormattedDate';
import { Resource } from 'components/frontend';
import { find } from 'lodash';
import lh from 'helpers/linkHandler';

class ResourceCard extends Component {

  static displayName = "Resource.Card";

  static propTypes = {
    history: PropTypes.object.isRequired,
    resource: PropTypes.object,
    context: PropTypes.object
  };

  constructor() {
    super();
    this.state = {
      infoHover: false
    };
    this.handlePreviewClick = this.handlePreviewClick.bind(this);
    this.handleInfoMouseOver = this.handleInfoMouseOver.bind(this);
    this.handleInfoMouseOut = this.handleInfoMouseOut.bind(this);
    this.handleInfoClick = this.handleInfoClick.bind(this);
    this.handleTagHover = this.handleTagHover.bind(this);
  }

  getResourceType(type) {
    let formattedType = type.toLowerCase().charAt(0).toUpperCase() + type.slice(1);
    if (type.toLowerCase() === 'pdf') {
      formattedType = 'PDF';
    }
    return formattedType;
  }

  getPreviewText(attr) {
    const type = attr.kind;
    const text = (attr.downloadable ?
      this.renderDownloadablePreview(type)
      : <span>
          {'View'}
          <i className="manicon manicon-arrow-right"></i>
        </span>
    );
    return text;
  }

  getCollectionResourceId(resource) {
    if (!resource) return null;
    const collectionResources = resource.relationships.collectionResources;
    const out = find(collectionResources, (obj) => {
      return obj.attributes.collectionId === this.props.context.id;
    });
    if (!out) return null;
    return out.id;
  }

  detailUrl() {
    const context = this.props.context;
    if (context.type === "collections") {
      const pid = context.attributes.projectId;
      const cid = context.id;
      const crid = this.getCollectionResourceId(this.props.resource);
      return lh.link("frontendProjectCollectionCollectionResource", pid, cid, crid);
    }
    if (context.type === "projects") {
      const resource = this.resource();
      const pid = context.id;
      const rid = resource.id;
      return lh.link("frontendProjectResource", pid, rid);
    }
  }

  handlePreviewClick(event) {
    event.preventDefault();
    const attr = this.resource().attributes;
    let action = null;
    switch (attr.kind.toLowerCase()) {
      case "link":
        action = window.open(attr.externalUrl);
        break;
      default:
        action = attr.downloadable ?
          window.open(attr.attachmentStyles.original)
          : this.handleInfoClick();
        break;
    }
    return action;
  }

  handleInfoMouseOver() {
    this.setState({
      infoHover: true
    });
  }

  handleInfoMouseOut() {
    this.setState({
      infoHover: false
    });
  }

  handleInfoClick() {
    this.props.history.push(this.detailUrl());
  }

  handleTagHover(event) {
    event.stopPropagation();
    this.setState({
      infoHover: false
    });
  }

  handleTagClick(event) {
    // Placeholder method, ultimately this will link
    // to the tag detail

    // Event handler, has access to react synthetic click event
    // Will need to be bound to the component in the constructor to use any component
    // state or props
    event.stopPropagation();
  }

  resource() {
    if (this.props.resource.type === "collectionResources") {
      return this.props.resource.relationships.resource;
    }
    return this.props.resource;
  }

  renderDownloadablePreview(type) {
    if (!type) return null;
    let out = null;
    switch (type.toLowerCase()) {
      case 'image':
      case 'interactive':
        out = (
          <span>
            {'Preview'}
            <i className="manicon manicon-eye-outline"></i>
          </span>
        );
        break;
      case 'link':
        out = (
          <span>
            {'Visit'}
            <i className="manicon manicon-arrow-right"></i>
          </span>
        );
        break;
      case 'video':
        out = (
          <span>
            {'Play'}
            <i className="manicon manicon-triangle-right"></i>
          </span>
        );
        break;
      default:
        out = (
          <span>
            {'Download'}
            <i className="manicon manicon-arrow-down"></i>
          </span>
        );
    }
    return out;
  }

  renderTags(resource) {
    if (!resource.attributes.tags) {
      return false;
    }

    function commaSeparate(index) {
      if (index >= resource.attributes.tags.length - 1) return false;
      return (
          <span>
          {', '}
        </span>
      );
    }

    return (
      <nav className="resource-tags">
        <ul>
          {resource.attributes.tags.map((tag, index) => {
            return (
                <div
                  key={index}
                  className="tag-link"
                  onMouseOver={this.handleTagHover}
                  onClick={this.handleTagClick}
                >
                  {tag}{commaSeparate(index)}
                </div>
            );
          })}
        </ul>
      </nav>
    );
  }

  render() {
    const resource = this.resource();
    if (!resource) return null;
    const attr = resource.attributes;

    const linkClass = classNames({
      thumbnail: true,
      'bg-image': attr.attachmentStyles.smallPortrait
    });

    const infoClass = classNames({
      'resource-info': true,
      hover: this.state.infoHover
    });


    return (
      <li className="resource-card">
        <div
          className="resource-link"
          onClick={this.handlePreviewClick}
        >
          <Resource.Thumbnail
            resource={resource}
          />
          <div className="preview-text">
            {this.getPreviewText(attr)}
          </div>
        </div>
        <section
          className={infoClass}
          onMouseOver={this.handleInfoMouseOver}
          onMouseOut={this.handleInfoMouseOut}
          onClick={this.handleInfoClick}
        >
          <div>
            <header
              className="resource-title"
            >
              <h4 dangerouslySetInnerHTML={{ __html: attr.titleFormatted }} />
            </header>
            <span className="resource-date">
              <FormattedDate
                format="MMMM, YYYY"
                date={attr.createdAt}
              />
            </span>
            <div
              to={this.detailUrl()}
              className="arrow-link"
            >
              <i className="manicon manicon-arrow-right"></i>
            </div>
          </div>

          {this.renderTags(resource)}
        </section>
      </li>
    );
  }
}

export default withRouter(ResourceCard);
