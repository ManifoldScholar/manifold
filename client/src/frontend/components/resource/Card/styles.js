import styled from "@emotion/styled";
import IconComposer from "global/components/utility/IconComposer";
import {
  respond,
  defaultTransitionProps,
  subtitlePrimary
} from "theme/styles/mixins";

export const Card = styled.article`
  --Thumbnail-Icon-background-color: transparent;

  display: flex;
`;

export const Preview = styled.div`
  display: flex;
  width: 37.8%;

  ${respond(`width: 155px;`, 40)}

  ${respond(`width: 135px;`, 75)}

  ${respond(`width: 155px;`, 85)}
`;

export const InfoLink = styled.a`
  position: relative;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  width: 62.2%;
  padding: 12px 15px;
  text-decoration: none;
  background-color: var(--color-base-neutral05);

  ${respond(`width: calc(100% - 155px);`, 40)}

  ${respond(`width: calc(100% - 135px);`, 75)}

  ${respond(`width: calc(100% - 155px);`, 85)}

  &--hover,
  &:focus-visible {
    cursor: pointer;
    outline: 0;
  }
`;

export const Title = styled.h4`
  font-family: var(--font-family-heading);
  margin: 0;
  margin-block-end: 6px;
  font-size: 16px;
  font-weight: var(--font-weight-regular);
  hyphens: none;

  ${respond(`font-size: 17px;`, 40)}
`;

export const Date = styled.span`
  ${subtitlePrimary}
  font-size: 12px;

  ${respond(`font-size: 14px;`, 40)}
`;

export const ArrowWrapper = styled.div`
  position: absolute;
  top: 50%;
  right: -10px;
  display: block;
  width: 26px;
  height: 26px;
  margin-block-start: -13px;
  font-size: 12px;
  text-align: center;
  text-decoration: none;
  background-color: var(--color-base-neutral20);
  border-radius: 100%;
  transition: color ${defaultTransitionProps},
    background-color ${defaultTransitionProps};

  ${respond(
    `right: -17.5px;
    width: 35px;
    height: 35px;
    margin-block-start: -17.5px;
    font-size: 15px;`,
    40
  )}

  &:hover,
  ${InfoLink}:hover &,
  ${InfoLink}:focus-visible & {
    color: var(--color-base-neutral-white);
    background-color: var(--hover-color);
  }
`;

export const ArrowIcon = styled(IconComposer)`
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
`;
