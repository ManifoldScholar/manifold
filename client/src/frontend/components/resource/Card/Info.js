import React from "react";
import PropTypes from "prop-types";
import FormattedDate from "global/components/FormattedDate";
import TagList from "../TagList";
import * as Styled from "./styles";

function Info({ resource, detailUrl, itemHeadingLevel = 4 }) {
  const attr = resource.attributes;
  const titleTag = `h${itemHeadingLevel}`;

  return (
    <Styled.InfoLink href={detailUrl}>
      <div>
        <header>
          <Styled.Title
            as={titleTag}
            dangerouslySetInnerHTML={{ __html: attr.titleFormatted }}
          />
        </header>
        <Styled.Date>
          Uploaded <FormattedDate format="MMMM, yyyy" date={attr.createdAt} />
        </Styled.Date>
        <Styled.ArrowWrapper>
          <Styled.ArrowIcon icon="arrowRight16" size={20} />
        </Styled.ArrowWrapper>
      </div>
      <TagList resource={resource} isCard disabledLinks />
    </Styled.InfoLink>
  );
}

Info.displayName = "ResourceCard.Info";

Info.propTypes = {
  resource: PropTypes.object.isRequired,
  detailUrl: PropTypes.string,
  itemHeadingLevel: PropTypes.oneOf([2, 3, 4, 5, 6])
};

export default Info;
