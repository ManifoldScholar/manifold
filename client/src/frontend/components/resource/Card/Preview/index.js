import React from "react";
import PropTypes from "prop-types";
import Resourceish from "frontend/components/resourceish";
import ResourcePreview from "frontend/components/resource-preview";
import Text from "./Text";
import * as Styled from "./styles";

function Preview({ resource, detailUrl }) {
  const attr = resource.attributes;
  const previewable = ResourcePreview.canPreview(resource);
  const linkable = attr.kind.toLowerCase() === "link";
  const downloadable = attr.downloadable;

  const PreviewInner = () => (
    <Styled.Inner>
      <Resourceish.Thumbnail resourceish={resource} />
      <Styled.TextWrapper>
        <Text resource={resource} />
      </Styled.TextWrapper>
    </Styled.Inner>
  );

  if (previewable)
    return (
      <ResourcePreview resource={resource} toggleType="card">
        <PreviewInner />
      </ResourcePreview>
    );

  if (linkable)
    return (
      <Styled.Link
        href={attr.externalUrl}
        target="_blank"
        rel="noopener noreferrer"
      >
        <PreviewInner />
      </Styled.Link>
    );

  if (downloadable)
    return (
      <Styled.Link href={attr.attachmentStyles.original} download={attr.slug}>
        <PreviewInner />
      </Styled.Link>
    );

  return (
    <Styled.Link href={detailUrl} tabIndex={-1}>
      <PreviewInner />
    </Styled.Link>
  );
}

Preview.displayName = "ResourceCard.Preview";

Preview.propTypes = {
  resource: PropTypes.object.isRequired,
  detailUrl: PropTypes.string
};

export default Preview;
