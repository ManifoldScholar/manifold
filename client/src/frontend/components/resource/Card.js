import React, { Component } from "react";
import PropTypes from "prop-types";
import { withRouter } from "react-router-dom";
import classNames from "classnames";
import FormattedDate from "global/components/FormattedDate";
import Resourceish from "frontend/components/resourceish";
import Preview from "frontend/components/resource-preview";
import TagList from "./TagList";

import lh from "helpers/linkHandler";

class ResourceCard extends Component {
  static displayName = "Card";

  static propTypes = {
    history: PropTypes.object.isRequired,
    resource: PropTypes.object,
    project: PropTypes.object.isRequired,
    resourceCollection: PropTypes.object
  };

  constructor() {
    super();
    this.state = {
      infoHover: false
    };
  }

  getResourceType(type) {
    let formattedType =
      type
        .toLowerCase()
        .charAt(0)
        .toUpperCase() + type.slice(1);
    if (type.toLowerCase() === "pdf") {
      formattedType = "PDF";
    }
    return formattedType;
  }

  getPreviewText(attr) {
    const { kind: type } = attr;
    const text = attr.downloadable ? (
      this.renderDownloadablePreview(type)
    ) : (
      <span>
        {"View"}
        <i className="manicon manicon-arrow-right" aria-hidden="true" />
      </span>
    );
    return text;
  }

  detailUrl() {
    if (this.props.resourceCollection) {
      return lh.link(
        "frontendProjectCollectionResource",
        this.props.project.attributes.slug,
        this.props.resourceCollection.attributes.slug,
        this.props.resource.attributes.slug
      );
    }
    return lh.link(
      "frontendProjectResource",
      this.props.project.attributes.slug,
      this.props.resource.attributes.slug
    );
  }

  previewable(resource) {
    return Preview.canPreview(resource);
  }

  linkable(resource) {
    return resource.attributes.kind.toLowerCase() === "link";
  }

  downloadable(resource) {
    return resource.attributes.downloadable || false;
  }

  doDownload(resource) {
    window.open(resource.attributes.attachmentStyles.original);
  }

  openLink(resource) {
    window.open(resource.attributes.externalUrl);
  }

  handlePreviewClick = event => {
    event.preventDefault();
    const resource = this.props.resource;
    if (this.previewable(resource)) return;
    if (this.downloadable(resource)) return this.doDownload(resource);
    if (this.linkable(resource)) return this.openLink(resource);
    // Open the resource detail view if all else fails.
    return this.handleInfoClick();
  };

  handleInfoMouseOver = () => {
    this.setState({
      infoHover: true
    });
  };

  handleInfoMouseOut = () => {
    this.setState({
      infoHover: false
    });
  };

  handleInfoClick = () => {
    this.props.history.push(this.detailUrl());
  };

  renderDownloadablePreview(type) {
    if (!type) return null;
    let out = null;
    switch (type.toLowerCase()) {
      case "image":
      case "interactive":
        out = (
          <span>
            {"Preview"}
            <i className="manicon manicon-eye-outline" aria-hidden="true" />
          </span>
        );
        break;
      case "link":
        out = (
          <span>
            {"Visit"}
            <i className="manicon manicon-arrow-right" aria-hidden="true" />
          </span>
        );
        break;
      case "video":
        out = (
          <span>
            {"Play"}
            <i className="manicon manicon-triangle-right" aria-hidden="true" />
          </span>
        );
        break;
      default:
        out = (
          <span>
            {"Download"}
            <i className="manicon manicon-arrow-down" aria-hidden="true" />
          </span>
        );
    }
    return out;
  }

  render() {
    const resource = this.props.resource;
    if (!resource) return null;
    const { attributes: attr } = resource;

    const infoClass = classNames({
      "resource-info": true,
      hover: this.state.infoHover
    });
    return (
      <li className="resource-card">
        <Preview resource={resource}>
          <div
            className="resource-link"
            onClick={this.handlePreviewClick}
            role="link"
            tabIndex="0"
          >
            <Resourceish.Thumbnail resourceish={resource} />
            <div className="preview-text">{this.getPreviewText(attr)}</div>
          </div>
        </Preview>
        {/* eslint-disable jsx-a11y/no-noninteractive-element-to-interactive-role */}
        <section
          className={infoClass}
          onMouseOver={this.handleInfoMouseOver}
          onMouseOut={this.handleInfoMouseOut}
          onClick={this.handleInfoClick}
          role="link"
          tabIndex="0"
        >
          {/* eslint-enable jsx-a11y/no-noninteractive-element-to-interactive-role */}
          <div>
            <header className="resource-title">
              <h4 dangerouslySetInnerHTML={{ __html: attr.titleFormatted }} />
            </header>
            <span className="resource-date">
              Uploaded{" "}
              <FormattedDate format="MMMM, YYYY" date={attr.createdAt} />
            </span>
            <div to={this.detailUrl()} className="arrow-link">
              <i className="manicon manicon-arrow-right" aria-hidden="true" />
            </div>
          </div>
          <TagList resource={resource} />
        </section>
      </li>
    );
  }
}

export default withRouter(ResourceCard);
