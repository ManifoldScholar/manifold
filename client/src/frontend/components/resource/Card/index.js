import React from "react";
import PropTypes from "prop-types";
import lh from "helpers/linkHandler";
import Info from "./Info";
import Preview from "./Preview";
import * as Styled from "./styles";

function ResourceCard({
  resource,
  resourceCollection,
  project,
  itemHeadingLevel = 4
}) {
  if (!resource) return null;

  function getDetailUrl() {
    if (resourceCollection) {
      return lh.link(
        "frontendProjectCollectionResource",
        project.attributes.slug,
        resourceCollection.attributes.slug,
        resource.attributes.slug
      );
    }
    return lh.link(
      "frontendProjectResource",
      project.attributes.slug,
      resource.attributes.slug
    );
  }

  return (
    <Styled.Card>
      <Styled.Preview>
        <Preview resource={resource} detailUrl={getDetailUrl()} />
      </Styled.Preview>
      <Info
        resource={resource}
        detailUrl={getDetailUrl()}
        itemHeadingLevel={itemHeadingLevel}
      />
    </Styled.Card>
  );
}

ResourceCard.propTypes = {
  resource: PropTypes.object,
  project: PropTypes.object.isRequired,
  resourceCollection: PropTypes.object,
  itemHeadingLevel: PropTypes.oneOf([2, 3, 4, 5, 6])
};

export default ResourceCard;
