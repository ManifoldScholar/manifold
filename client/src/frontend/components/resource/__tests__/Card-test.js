import Card from "../Card";

describe("frontend/components/resource/Card", () => {
  def("project", () => factory("project"));
  def("resource", () => factory("resource"));
  def("root", () => <Card resource={$resource} project={$project} />);

  it("matches the snapshot", () => {
    expect(shallow($root)).toMatchSnapshot();
  });
});
