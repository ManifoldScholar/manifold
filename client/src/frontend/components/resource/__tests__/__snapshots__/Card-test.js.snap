// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`frontend/components/resource/Card matches the snapshot 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
